File: bindings/csharp/En16931.SemanticJson/SemanticPath.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace En16931.SemanticJson;

/// <summary>Whether a term identifier names a business term or a business group.</summary>
public enum TermKind
{
    /// <summary>A business term, which carries a value.</summary>
    Bt,

    /// <summary>A business group, which carries terms and further groups.</summary>
    Bg,
}

/// <summary>
/// One step of a semantic path: a term identifier, or an occurrence index following one.
/// </summary>
/// <param name="Text">the segment as the path writes it</param>
/// <param name="IsIndex">whether the segment is an occurrence index</param>
public readonly record struct PathSegment(string Text, bool IsIndex)
{
    /// <summary>Returns the kind of term the segment names.</summary>
    /// <exception cref="InvalidOperationException">if the segment is an index</exception>
    public TermKind Kind => IsIndex
        ? throw new InvalidOperationException("an index segment names no term")
        : Text.StartsWith("BT-", StringComparison.Ordinal) ? TermKind.Bt : TermKind.Bg;

    /// <summary>
    /// Tells whether the segment names an extension term, which carries a namespace between
    /// the kind and the number (specification, section 5.6).
    /// </summary>
    public bool IsExtension
    {
        get
        {
            if (IsIndex)
            {
                return false;
            }

            string rest = Text.Substring(3);
            return rest.Length > 0 && char.IsAsciiLetterUpper(rest[0]);
        }
    }
}

/// <summary>
/// A semantic path: the address of a value, or of a business group instance, as the
/// specification, section 5.1 writes it.
/// </summary>
public sealed class SemanticPath : IComparable<SemanticPath>, IEquatable<SemanticPath>
{
    private static readonly SemanticPath RootPath = new(string.Empty, Array.Empty<PathSegment>());

    private readonly PathSegment[] _segments;

    private SemanticPath(string text, PathSegment[] segments)
    {
        Text = text;
        _segments = segments;
    }

    /// <summary>Returns the path of the document itself, which has no segment.</summary>
    /// <returns>the root</returns>
    public static SemanticPath Root() => RootPath;

    /// <summary>Returns the path as the document writes it.</summary>
    public string Text { get; }

    /// <summary>Returns the segments of the path, outermost first.</summary>
    public IReadOnlyList<PathSegment> Segments => _segments;

    /// <summary>Tells whether this is the document itself.</summary>
    public bool IsRoot => _segments.Length == 0;

    /// <summary>Returns the identifier of the term the path ends at.</summary>
    /// <exception cref="InvalidOperationException">if this is the root</exception>
    public string Term
    {
        get
        {
            for (int at = _segments.Length - 1; at >= 0; at--)
            {
                if (!_segments[at].IsIndex)
                {
                    return _segments[at].Text;
                }
            }

            throw new InvalidOperationException("the root of a document names no term");
        }
    }

    /// <summary>Returns the term identifiers of the path, outermost first.</summary>
    /// <returns>the chain of identifiers, without the occurrence indices</returns>
    public IReadOnlyList<string> TermIds()
    {
        List<string> ids = new(_segments.Length);
        foreach (PathSegment segment in _segments)
        {
            if (!segment.IsIndex)
            {
                ids.Add(segment.Text);
            }
        }

        return ids;
    }

    /// <summary>
    /// Reads a path and checks it against the grammar of the specification, section 5.1: a
    /// chain of business group steps ending at a business term, each step optionally followed
    /// by one occurrence index.
    /// </summary>
    /// <param name="text">the path</param>
    /// <param name="path">the path, where the text is one</param>
    /// <returns>whether the text satisfies the grammar</returns>
    public static bool TryParse(string text, out SemanticPath path)
    {
        if (!TryParseAny(text, out path) || path.IsRoot)
        {
            path = RootPath;
            return false;
        }

        IReadOnlyList<PathSegment> segments = path.Segments;
        for (int at = 0; at < segments.Count; at++)
        {
            if (segments[at].IsIndex)
            {
                continue;
            }

            bool last = at == segments.Count - 1
                || (at == segments.Count - 2 && segments[at + 1].IsIndex);
            if (last != (segments[at].Kind == TermKind.Bt))
            {
                path = RootPath;
                return false;
            }
        }

        return true;
    }

    /// <summary>Reads a path, or a business group instance path, without checking a kind.</summary>
    /// <param name="text">the path</param>
    /// <returns>the path</returns>
    /// <exception cref="EsjFormatException">if the text is not a path</exception>
    public static SemanticPath Parse(string text)
    {
        if (text == "/" || text.Length == 0)
        {
            return RootPath;
        }

        if (!TryParseAny(text, out SemanticPath path))
        {
            throw new EsjFormatException("ESJ-L1-PATH-SYNTAX", text + " is not a semantic path");
        }

        return path;
    }

    /// <summary>
    /// Reads a path that may end at a business group, which is what a group instance is.
    /// </summary>
    /// <param name="text">the path</param>
    /// <param name="path">the path, where the text is one</param>
    /// <returns>whether the text is a chain of segments</returns>
    public static bool TryParseAny(string text, out SemanticPath path)
    {
        path = RootPath;
        if (text.Length == 0 || text == "/")
        {
            return true;
        }

        if (text[0] != '/')
        {
            return false;
        }

        string[] tokens = text.Substring(1).Split('/');
        List<PathSegment> segments = new(tokens.Length);
        bool previousWasIndex = true;
        foreach (string token in tokens)
        {
            if (token.Length == 0)
            {
                return false;
            }

            if (char.IsAsciiDigit(token[0]))
            {
                if (previousWasIndex || !IsIndexToken(token))
                {
                    return false;
                }

                segments.Add(new PathSegment(token, true));
                previousWasIndex = true;
                continue;
            }

            if (!IsTermToken(token))
            {
                return false;
            }

            segments.Add(new PathSegment(token, false));
            previousWasIndex = false;
        }

        path = new SemanticPath(text, segments.ToArray());
        return true;
    }

    private static bool IsIndexToken(string token) =>
        token == "0" || (token[0] != '0' && AllDigits(token));

    private static bool AllDigits(string token)
    {
        foreach (char character in token)
        {
            if (!char.IsAsciiDigit(character))
            {
                return false;
            }
        }

        return true;
    }

    private static bool IsTermToken(string token)
    {
        if (!token.StartsWith("BT-", StringComparison.Ordinal)
            && !token.StartsWith("BG-", StringComparison.Ordinal))
        {
            return false;
        }

        string rest = token.Substring(3);
        if (rest.Length == 0)
        {
            return false;
        }

        if (char.IsAsciiDigit(rest[0]))
        {
            return rest[0] != '0' && AllDigits(rest);
        }

        int dash = rest.IndexOf('-', StringComparison.Ordinal);
        if (dash <= 0 || dash == rest.Length - 1)
        {
            return false;
        }

        string space = rest.Substring(0, dash);
        if (!char.IsAsciiLetterUpper(space[0]))
        {
            return false;
        }

        foreach (char character in space)
        {
            if (!char.IsAsciiLetterUpper(character) && !char.IsAsciiDigit(character))
            {
                return false;
            }
        }

        return AllDigits(rest.Substring(dash + 1));
    }

    /// <summary>
    /// Tells whether this path lies at or below another one, which is what makes the subtree
    /// of a business group instance a range of the canonical order.
    /// </summary>
    /// <param name="prefix">the path the subtree hangs under</param>
    /// <returns>whether every segment of the prefix opens this path</returns>
    public bool StartsWith(SemanticPath prefix)
    {
        ArgumentNullException.ThrowIfNull(prefix);
        if (prefix._segments.Length > _segments.Length)
        {
            return false;
        }

        for (int at = 0; at < prefix._segments.Length; at++)
        {
            if (!string.Equals(prefix._segments[at].Text, _segments[at].Text, StringComparison.Ordinal)
                || prefix._segments[at].IsIndex != _segments[at].IsIndex)
            {
                return false;
            }
        }

        return true;
    }

    /// <summary>Returns the path made of the first segments of this one.</summary>
    /// <param name="count">how many segments to keep</param>
    /// <returns>the prefix</returns>
    public SemanticPath Prefix(int count)
    {
        if (count >= _segments.Length)
        {
            return this;
        }

        PathSegment[] segments = new PathSegment[count];
        Array.Copy(_segments, segments, count);
        StringBuilder text = new();
        foreach (PathSegment segment in segments)
        {
            text.Append('/').Append(segment.Text);
        }

        return new SemanticPath(text.ToString(), segments);
    }

    /// <summary>
    /// Compares two paths in the canonical path order of the specification, section 7.4.
    /// </summary>
    /// <param name="other">the path to compare with</param>
    /// <returns>a negative number, zero or a positive number</returns>
    public int CompareTo(SemanticPath? other)
    {
        if (other is null)
        {
            return 1;
        }

        int common = Math.Min(_segments.Length, other._segments.Length);
        for (int at = 0; at < common; at++)
        {
            int order = Compare(_segments[at], other._segments[at]);
            if (order != 0)
            {
                return order;
            }
        }

        return _segments.Length.CompareTo(other._segments.Length);
    }

    private static int Compare(PathSegment left, PathSegment right)
    {
        if (left.IsIndex != right.IsIndex)
        {
            return left.IsIndex ? 1 : -1;
        }

        if (left.IsIndex)
        {
            return CompareDigits(left.Text, right.Text);
        }

        int kinds = left.Kind.CompareTo(right.Kind);
        if (kinds != 0)
        {
            return kinds;
        }

        string leftSpace = NamespaceOf(left.Text);
        string rightSpace = NamespaceOf(right.Text);
        int spaces = string.CompareOrdinal(leftSpace, rightSpace);
        if (spaces != 0)
        {
            return leftSpace.Length == 0 ? -1 : rightSpace.Length == 0 ? 1 : spaces;
        }

        string leftNumber = NumberOf(left.Text);
        string rightNumber = NumberOf(right.Text);
        int numbers = CompareDigits(leftNumber, rightNumber);
        return numbers != 0 ? numbers : string.CompareOrdinal(left.Text, right.Text);
    }

    private static string NamespaceOf(string term)
    {
        string rest = term.Substring(3);
        int dash = rest.IndexOf('-', StringComparison.Ordinal);
        return dash < 0 ? string.Empty : rest.Substring(0, dash);
    }

    private static string NumberOf(string term)
    {
        string rest = term.Substring(3);
        int dash = rest.IndexOf('-', StringComparison.Ordinal);
        return dash < 0 ? rest : rest.Substring(dash + 1);
    }

    /// <summary>
    /// Compares two digit strings numerically, without converting either to a machine
    /// integer (specification, section 7.4).
    /// </summary>
    /// <param name="left">the first digit string</param>
    /// <param name="right">the second digit string</param>
    /// <returns>a negative number, zero or a positive number</returns>
    public static int CompareDigits(string left, string right)
    {
        string first = TrimZeros(left);
        string second = TrimZeros(right);
        return first.Length != second.Length
            ? first.Length.CompareTo(second.Length)
            : string.CompareOrdinal(first, second);
    }

    private static string TrimZeros(string digits)
    {
        int at = 0;
        while (at < digits.Length - 1 && digits[at] == '0')
        {
            at++;
        }

        return digits.Substring(at);
    }

    /// <inheritdoc />
    public bool Equals(SemanticPath? other) =>
        other is not null && string.Equals(Text, other.Text, StringComparison.Ordinal);

    /// <inheritdoc />
    public override bool Equals(object? obj) => Equals(obj as SemanticPath);

    /// <inheritdoc />
    public override int GetHashCode() => Text.GetHashCode(StringComparison.Ordinal);

    /// <inheritdoc />
    public override string ToString() => Text;

    /// <summary>The canonical path order of the specification, section 7.4, as a comparer.</summary>
    public static IComparer<SemanticPath> CanonicalOrder { get; } = Comparer<SemanticPath>.Create(
        (left, right) => left.CompareTo(right));
}

File: esj-core/src/main/java/de/bsnsoft/esj/SemanticPath.java
package de.bsnsoft.esj;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

/**
 * The absolute address of one business term occurrence, or of one business group
 * instance, inside a document (specification, section 5).
 *
 * <p>A <em>value path</em> ends at a business term and is what the member names of
 * {@code values} are: {@code /BG-25/0/BG-29/BT-146}. A <em>group path</em> ends at a
 * business group and names one group instance: {@code /BG-25/0}. The root of the document
 * is the path with no segments; its text is the empty string.
 *
 * <p>Instances are immutable. Two paths are equal exactly when their texts are equal,
 * code point by code point; there is no normalization and no case folding
 * (specification, section 5.5). The natural order is the canonical path order of the
 * specification, section 7.4, which needs no registry.
 */
public final class SemanticPath implements Comparable<SemanticPath> {

    private static final SemanticPath ROOT = new SemanticPath("", List.of());

    private static final Comparator<SemanticPath> CANONICAL_ORDER = SemanticPath::compareCanonically;

    private final String text;
    private final List<PathSegment> segments;

    private SemanticPath(String text, List<PathSegment> segments) {
        this.text = text;
        this.segments = segments;
    }

    /**
     * Returns the root of the document: the path with no segments, whose text is the
     * empty string. It is the parent of every path that has exactly one term step, and it
     * is the path of a finding that is about the document as a whole.
     *
     * @return the root path
     */
    public static SemanticPath root() {
        return ROOT;
    }

    /**
     * Parses a value path: a path that ends at a business term, which is what a member
     * name of {@code values} is.
     *
     * @param text the path text, starting with a solidus
     * @return the parsed path
     * @throws EsjFormatException   if the text does not match the path grammar of the
     *                              specification, section 5.1, or does not end at a
     *                              business term
     * @throws NullPointerException if {@code text} is {@code null}
     */
    public static SemanticPath of(String text) {
        return parse(text, TermKind.BT);
    }

    /**
     * Parses a group path: a path that ends at a business group and therefore names one
     * group instance.
     *
     * @param text the path text, starting with a solidus
     * @return the parsed path
     * @throws EsjFormatException   if the text does not match the path grammar of the
     *                              specification, section 5.1, or does not end at a
     *                              business group
     * @throws NullPointerException if {@code text} is {@code null}
     */
    public static SemanticPath group(String text) {
        return parse(text, TermKind.BG);
    }

    /**
     * Builds a path from segments. The segments must form the grammar of the
     * specification, section 5.1: term steps, each optionally followed by one index
     * segment, with a business term only in the last step.
     *
     * @param segments the segments, from the root down
     * @return the path
     * @throws EsjFormatException   if the segments do not form a path
     * @throws NullPointerException if {@code segments} or an element is {@code null}
     */
    public static SemanticPath ofSegments(List<PathSegment> segments) {
        Objects.requireNonNull(segments, "segments");
        List<PathSegment> copy = List.copyOf(segments);
        if (copy.isEmpty()) {
            return ROOT;
        }
        checkShape(copy);
        return new SemanticPath(join(copy), copy);
    }

    /**
     * Returns the comparator that implements the canonical path order of the
     * specification, section 7.4. It compares segment sequences rather than strings, it
     * needs no registry, and it is a total order on every path this class can hold.
     *
     * @return the canonical path order
     */
    public static Comparator<SemanticPath> canonicalOrder() {
        return CANONICAL_ORDER;
    }

    /**
     * Returns the segments of this path, from the root down.
     *
     * @return an unmodifiable list of segments, empty for the root
     */
    public List<PathSegment> segments() {
        return segments;
    }

    /**
     * Tells whether this path is the root of the document.
     *
     * @return {@code true} if the path has no segments
     */
    public boolean isRoot() {
        return segments.isEmpty();
    }

    /**
     * Tells whether this path addresses a business term and can therefore carry a value.
     *
     * @return {@code true} if the last term segment is a business term
     */
    public boolean isValuePath() {
        return !isRoot() && lastTerm().kind() == TermKind.BT;
    }

    /**
     * Tells whether this path addresses a business group instance.
     *
     * @return {@code true} if the last term segment is a business group
     */
    public boolean isGroupPath() {
        return !isRoot() && lastTerm().kind() == TermKind.BG;
    }

    /**
     * Returns the enclosing group instance, or the root when this path has a single term
     * step.
     *
     * @return the parent path
     * @throws IllegalStateException if this path is the root, which has no parent
     */
    public SemanticPath parent() {
        if (isRoot()) {
            throw new IllegalStateException("the root path has no parent");
        }
        int cut = segments.size() - (lastSegmentIsIndex() ? 2 : 1);
        return prefix(cut);
    }

    /**
     * Returns the group instances this path lies in, from the outermost to the innermost.
     * For {@code /BG-25/0/BG-29/BT-146} these are {@code /BG-25/0} and
     * {@code /BG-25/0/BG-29}.
     *
     * @return an unmodifiable list of group paths, empty for a path that lies at the root
     */
    public List<SemanticPath> groupPaths() {
        List<SemanticPath> result = new ArrayList<>();
        for (int i = 0; i < segments.size(); i++) {
            if (segments.get(i) instanceof PathSegment.Term term && term.kind() == TermKind.BG) {
                int end = i + 1;
                if (end < segments.size() && segments.get(end) instanceof PathSegment.Index) {
                    end++;
                }
                result.add(prefix(end));
            }
        }
        return List.copyOf(result);
    }

    /**
     * Returns the identifier of the last term of this path, for example {@code BT-146}.
     *
     * @return the term identifier
     * @throws IllegalStateException if this path is the root, which names no term
     */
    public String term() {
        if (isRoot()) {
            throw new IllegalStateException("the root path names no term");
        }
        return lastTerm().id();
    }

    /**
     * Returns the term segment of the last step of this path.
     *
     * @return the last term segment
     * @throws IllegalStateException if this path is the root, which names no term
     */
    public PathSegment.Term termSegment() {
        if (isRoot()) {
            throw new IllegalStateException("the root path names no term");
        }
        return lastTerm();
    }

    /**
     * Returns the occurrence index of the last step of this path, if it carries one.
     *
     * @return the index segment of the last step, or an empty optional
     */
    public Optional<PathSegment.Index> index() {
        if (lastSegmentIsIndex()) {
            return Optional.of((PathSegment.Index) segments.get(segments.size() - 1));
        }
        return Optional.empty();
    }

    /**
     * Tells whether the last term of this path belongs to an extension rather than to the
     * core model.
     *
     * @return {@code true} if the last term segment carries a namespace
     * @throws IllegalStateException if this path is the root, which names no term
     */
    public boolean isExtension() {
        if (isRoot()) {
            throw new IllegalStateException("the root path names no term");
        }
        return lastTerm().isExtension();
    }

    /**
     * Tells whether any term of this path belongs to an extension.
     *
     * @return {@code true} if at least one term segment carries a namespace
     */
    public boolean hasExtensionSegment() {
        for (PathSegment segment : segments) {
            if (segment instanceof PathSegment.Term term && term.isExtension()) {
                return true;
            }
        }
        return false;
    }

    /**
     * Returns the identifiers of the terms of this path, from the root down.
     *
     * @return an unmodifiable list of term identifiers, empty for the root
     */
    public List<String> termIds() {
        List<String> ids = new ArrayList<>(segments.size());
        for (PathSegment segment : segments) {
            if (segment instanceof PathSegment.Term term) {
                ids.add(term.id());
            }
        }
        return List.copyOf(ids);
    }

    /**
     * Tells whether this path lies at or below another path.
     *
     * @param other the candidate prefix
     * @return {@code true} if the segments of {@code other} are a prefix of the segments
     *         of this path
     * @throws NullPointerException if {@code other} is {@code null}
     */
    public boolean startsWith(SemanticPath other) {
        Objects.requireNonNull(other, "other");
        if (other.segments.size() > segments.size()) {
            return false;
        }
        return segments.subList(0, other.segments.size()).equals(other.segments);
    }

    /**
     * Returns the path formed by the first {@code count} segments of this path.
     *
     * @param count the number of segments to keep
     * @return the prefix path
     * @throws IllegalArgumentException if {@code count} is negative or greater than the
     *                                  number of segments
     */
    public SemanticPath prefix(int count) {
        if (count < 0 || count > segments.size()) {
            throw new IllegalArgumentException("segment count out of range: " + count);
        }
        if (count == segments.size()) {
            return this;
        }
        if (count == 0) {
            return ROOT;
        }
        List<PathSegment> head = List.copyOf(segments.subList(0, count));
        return new SemanticPath(join(head), head);
    }

    @Override
    public int compareTo(SemanticPath other) {
        return compareCanonically(this, other);
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof SemanticPath other && text.equals(other.text);
    }

    @Override
    public int hashCode() {
        return text.hashCode();
    }

    /**
     * Returns the path text, which is the empty string for the root.
     *
     * @return the path as it appears in a document
     */
    @Override
    public String toString() {
        return text;
    }

    private PathSegment.Term lastTerm() {
        PathSegment last = segments.get(segments.size() - 1);
        if (last instanceof PathSegment.Term term) {
            return term;
        }
        return (PathSegment.Term) segments.get(segments.size() - 2);
    }

    private boolean lastSegmentIsIndex() {
        return !segments.isEmpty() && segments.get(segments.size() - 1) instanceof PathSegment.Index;
    }

    private static SemanticPath parse(String text, TermKind expectedLastKind) {
        Objects.requireNonNull(text, "text");
        if (text.isEmpty() || text.charAt(0) != '/') {
            throw new EsjFormatException("a semantic path starts with a solidus: " + text);
        }
        List<PathSegment> segments = new ArrayList<>();
        int position = 1;
        TermKind lastKind = null;
        while (position <= text.length()) {
            int end = text.indexOf('/', position);
            if (end < 0) {
                end = text.length();
            }
            String token = text.substring(position, end);
            if (token.isEmpty()) {
                throw new EsjFormatException("empty segment in semantic path: " + text);
            }
            if (PathSegment.isDigits(token)) {
                if (segments.isEmpty() || segments.get(segments.size() - 1) instanceof PathSegment.Index) {
                    throw new EsjFormatException(
                            "an occurrence index follows a term segment: " + text);
                }
                segments.add(new PathSegment.Index(token));
            } else {
                if (lastKind == TermKind.BT) {
                    throw new EsjFormatException(
                            "only the last step of a semantic path names a business term: " + text);
                }
                PathSegment.Term term = parseTerm(token, text);
                lastKind = term.kind();
                segments.add(term);
            }
            position = end + 1;
        }
        if (lastKind != expectedLastKind) {
            throw new EsjFormatException(expectedLastKind == TermKind.BT
                    ? "a value path ends at a business term: " + text
                    : "a group path ends at a business group: " + text);
        }
        return new SemanticPath(text, List.copyOf(segments));
    }

    private static PathSegment.Term parseTerm(String token, String path) {
        if (token.length() < 4 || token.charAt(2) != '-') {
            throw new EsjFormatException("not a term segment: " + token + " in " + path);
        }
        TermKind kind;
        try {
            kind = TermKind.fromPrefix(token.substring(0, 2));
        } catch (EsjFormatException e) {
            throw new EsjFormatException("not a term segment: " + token + " in " + path, e);
        }
        String rest = token.substring(3);
        int separator = rest.indexOf('-');
        try {
            if (separator < 0) {
                return new PathSegment.Term(kind, null, rest);
            }
            return new PathSegment.Term(kind, rest.substring(0, separator), rest.substring(separator + 1));
        } catch (EsjFormatException e) {
            throw new EsjFormatException("not a term segment: " + token + " in " + path, e);
        }
    }

    private static void checkShape(List<PathSegment> segments) {
        TermKind lastKind = null;
        for (int i = 0; i < segments.size(); i++) {
            PathSegment segment = Objects.requireNonNull(segments.get(i), "segment");
            if (segment instanceof PathSegment.Term term) {
                if (lastKind == TermKind.BT) {
                    throw new EsjFormatException(
                            "only the last step of a semantic path names a business term");
                }
                lastKind = term.kind();
            } else if (i == 0 || segments.get(i - 1) instanceof PathSegment.Index) {
                throw new EsjFormatException("an occurrence index follows a term segment");
            }
        }
    }

    private static String join(List<PathSegment> segments) {
        StringBuilder builder = new StringBuilder();
        for (PathSegment segment : segments) {
            builder.append('/').append(segment.text());
        }
        return builder.toString();
    }

    private static int compareCanonically(SemanticPath left, SemanticPath right) {
        List<PathSegment> a = left.segments;
        List<PathSegment> b = right.segments;
        int common = Math.min(a.size(), b.size());
        for (int i = 0; i < common; i++) {
            int result = compareSegments(a.get(i), b.get(i));
            if (result != 0) {
                return result;
            }
        }
        return Integer.compare(a.size(), b.size());
    }

    private static int compareSegments(PathSegment left, PathSegment right) {
        if (left instanceof PathSegment.Term a) {
            if (!(right instanceof PathSegment.Term b)) {
                return -1;
            }
            int result = Integer.compare(a.kind().canonicalRank(), b.kind().canonicalRank());
            if (result != 0) {
                return result;
            }
            result = compareNamespaces(a.namespace(), b.namespace());
            if (result != 0) {
                return result;
            }
            result = compareDigits(a.number(), b.number());
            if (result != 0) {
                return result;
            }
            return CodePoints.compare(a.number(), b.number());
        }
        if (right instanceof PathSegment.Term) {
            return 1;
        }
        return compareDigits(((PathSegment.Index) left).digits(), ((PathSegment.Index) right).digits());
    }

    private static int compareNamespaces(String left, String right) {
        if (left == null) {
            return right == null ? 0 : -1;
        }
        if (right == null) {
            return 1;
        }
        return CodePoints.compare(left, right);
    }

    private static int compareDigits(String left, String right) {
        String a = stripLeadingZeros(left);
        String b = stripLeadingZeros(right);
        if (a.length() != b.length()) {
            return Integer.compare(a.length(), b.length());
        }
        return a.compareTo(b);
    }

    private static String stripLeadingZeros(String digits) {
        int i = 0;
        while (i < digits.length() - 1 && digits.charAt(i) == '0') {
            i++;
        }
        return digits.substring(i);
    }
}
